File: http/src/main/java/com/xinzy/http/HttpParam.java
package com.xinzy.http;

import android.text.TextUtils;
import android.util.ArrayMap;

import com.google.gson.Gson;

import java.io.File;
import java.net.FileNameMap;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import okhttp3.FormBody;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

/**
 * Created by Xinzy on 2017/5/19.
 */

class HttpParam {

    static final String ENCODE = "UTF-8";

    public static final MediaType MEDIA_TYPE_PLAIN = MediaType.parse("text/plain;charset=utf-8");
    public static final MediaType MEDIA_TYPE_JSON = MediaType.parse("application/json;charset=utf-8");
    public static final MediaType MEDIA_TYPE_STREAM = MediaType.parse("application/octet-stream");

    private static final int TYPE_DEFAULT = 0;
    private static final int TYPE_MULTI = 1;
    private static final int TYPE_JSON = 2;

    List<Entry> params = new ArrayList<>();;
    private int paramType = TYPE_DEFAULT;

    HttpParam add(String key, String val) {
        if (!TextUtils.isEmpty(key)) {
            val = val == null ? "" : val;
            params.add(new Entry(key, val));
        }
        return this;
    }

    HttpParam add(Map<String, String> vals) {
        if (vals != null && vals.size() > 0) {
            Set<String> keys = vals.keySet();
            for (String key : keys) {
                add(key, vals.get(key));
            }
        }
        return this;
    }

    HttpParam multi(String key, File file, String filename) {
        if (!TextUtils.isEmpty(key)) {
            paramType = TYPE_MULTI;
            params.add(new Entry(key, filename, file));
        }
        return this;
    }

    HttpParam merge(HttpParam param) {
        if (param != null && param.params != null && param.params.size() > 0) {
            params.addAll(param.params);
        }
        return this;
    }

    HttpParam asJson() {
        paramType = TYPE_JSON;
        return this;
    }

    RequestBody body() {
        if (params.isEmpty()) {
            return RequestBody.create(MEDIA_TYPE_PLAIN, new byte[0]);
        }
        if (paramType == TYPE_MULTI) {
            MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);
            for (Entry param : params) {
                if (param.isMulti) {
                    builder.addFormDataPart(param.key, param.value, RequestBody.create(param.getMediaType(), param.file));
                } else {
                    builder.addFormDataPart(param.key, Utils.encode(param.value));
                }
            }
            return builder.build();
        } else if (paramType == TYPE_JSON) {
            return RequestBody.create(MEDIA_TYPE_JSON, toJson());
        } else {
            FormBody.Builder builder = new FormBody.Builder();
            for (Entry param : params) {
                if (!param.isMulti) {
                    builder.add(param.key, param.value);
                }
            }
            return builder.build();
        }
    }

    Map<String, String> parameters() {
        Map<String, String> map = new ArrayMap<>();
        for (Entry entry : params) {
            if (entry != null && !entry.isMulti) {
                map.put(entry.key, entry.value);
            }
        }

        return map;
    }

    private String toJson() {
        if (params.size() == 0) {
            return "{}";
        } else {
            Map<String, String> map = new ArrayMap<>(params.size());
            for (Entry entry : params) {
                if (!entry.isMulti) {
                    map.put(entry.key, entry.value);
                }
            }
            return new Gson().toJson(map);
        }
    }

    static class Entry {
        String key;
        String value;
        File file;
        boolean isMulti;

        Entry(String key, String value) {
            this.key = key;
            this.value = value;
        }

        Entry(String key, String value, File file) {
            this.key = key;
            this.value = value;
            this.file = file;
            this.isMulti = true;
        }

        MediaType getMediaType() {
            FileNameMap fileNameMap = URLConnection.getFileNameMap();
            String name = file.getName().replace("#", "");
            String contentType = fileNameMap.getContentTypeFor(name);
            if (contentType != null) {
                return MediaType.parse(contentType);
            }
            return MEDIA_TYPE_STREAM;
        }
    }
}

File: httpkotlin/src/main/java/com/xinzy/http/kotlin/HttpParam.kt
package com.xinzy.http.kotlin

import android.text.TextUtils
import okhttp3.FormBody
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import org.json.JSONObject
import java.io.File
import java.net.URLConnection

/**
 * Created by Xinzy on 2018/10/12.
 */
internal class HttpParam {

    companion object {
        internal const val ENCODE = "UTF-8"

        private val MEDIA_TYPE_PLAIN = MediaType.parse("text/plain;charset=utf-8")
        private val MEDIA_TYPE_JSON = MediaType.parse("application/json;charset=utf-8")
        private val MEDIA_TYPE_STREAM = MediaType.parse("application/octet-stream")
    }

    val params = mutableListOf<Entry>()
    var isMulti = false
    var isJSON = false
    var mJSON: String? = null

    fun add(key: String, value: String?): HttpParam {
        if (!TextUtils.isEmpty(key)) {
            val v = value ?: ""
            params.add(Entry(key, v))
        }
        return this
    }

    fun add(vals: Map<String, String>?): HttpParam {
        if (vals != null && vals.isNotEmpty()) {
            vals.keys.forEach { add(it, vals[it])}
        }
        return this
    }

    fun asJson(): HttpParam {
        isJSON = true
        return this
    }

    fun json(json: String?): HttpParam {
        mJSON = json
        isJSON = true
        return this
    }

    fun multi(key: String, file: File, filename: String): HttpParam {
        if (!TextUtils.isEmpty(key)) {
            isMulti = true
            params.add(Entry(key, filename, file))
        }
        return this
    }

    fun merge(param: HttpParam?): HttpParam {
        if (param?.params != null && param.params.isNotEmpty()) {
            params.addAll(param.params)
        }
        return this
    }

    fun body(): RequestBody {
        if (isJSON) {
            return when {
                mJSON != null -> RequestBody.create(MEDIA_TYPE_JSON, mJSON!!)
                params.isEmpty() -> RequestBody.create(MEDIA_TYPE_JSON, "{}")
                else -> RequestBody.create(MEDIA_TYPE_JSON, toJSON())
            }
        }
        if (params.isEmpty()) {
            return RequestBody.create(MEDIA_TYPE_PLAIN, "")
        }

        return when {
            params.isEmpty() -> RequestBody.create(MEDIA_TYPE_PLAIN, "")
            isMulti -> {
                val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
                params.forEach {
                    if (it.isMulti) {
                        builder.addFormDataPart(it.key, it.value, RequestBody.create(it.mediaType, it.file!!))
                    } else {
                        builder.addFormDataPart(it.key, encode(it.value))
                    }
                }
                builder.build()
            }
            else -> {
                val builder = FormBody.Builder()
                params.filterNot { it.isMulti }.forEach { builder.add(it.key, encode(it.value)) }
                builder.build()
            }
        }
    }

    fun parameter(): Map<String, String> {
        return if (params.isEmpty()) {
            mapOf()
        } else {
            val map = mutableMapOf<String, String>()
            params.filterNot { !it.isMulti }.forEach { map[it.key] = it.value }
            map
        }
    }

    fun splitParam(): String {
        val sb = StringBuffer()
        return if (params.size > 0) {
            for (entry in params) {
                sb.append(entry.key).append("=").append(encode(entry.value)).append('&')
            }
            sb.substring(0, sb.length - 1)
        } else ""
    }

    private fun toJSON(): String {
        return try {
            JSONObject(parameter()).toString()
        } catch (e: Exception) {
            "{}"
        }
    }

    internal class Entry {
        var key: String
        var value: String
        var file: File? = null
        var isMulti: Boolean = false

        val mediaType: MediaType?
            get() {
                if (file == null) throw IllegalArgumentException("upload file is null")

                val fileNameMap = URLConnection.getFileNameMap()
                val name = file!!.name.replace("#", "")
                val contentType = fileNameMap.getContentTypeFor(name)
                return if (contentType != null) {
                    MediaType.parse(contentType)
                } else MEDIA_TYPE_STREAM
            }

        constructor(key: String, value: String) {
            this.key = key
            this.value = value
        }

        constructor(key: String, value: String, file: File) {
            this.key = key
            this.value = value
            this.file = file
            this.isMulti = true
        }
    }
}